src/slider/slider.tsx

```diff
@@ -0,0 +1,285 @@
+import React, {
+  Fragment,
+  KeyboardEvent,
+  ReactNode,
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState
+} from 'react';
+
+import classNames from 'classnames';
+
+import {isArray} from '../global/typescript-utils';
+
+import styles from './slider.css';
+import {
+  adjustValues,
+  calculateValue,
+  HUNDRED,
+  toPercent,
+  validateValue,
+  toRange,
+  calculateMarks
+} from './slider.utils';
+
+type Mark = {
+  value: number;
+  label?: ReactNode;
+}
+
+type Props = {
+  defaultValue?: number | number[];
+  value?: number | number[];
+  min?: number;
+  max?: number;
+  step?: number;
+  disabled?: boolean;
+  marks?: Mark[] | boolean
+  showTicks?: boolean;
+  showTag?: boolean;
+  className?: string;
+  renderTag?: (value: number) => ReactNode;
+  onChange?: (value: number | number[]) => void;
+};
+
+export const Slider: React.FC<Props> = ({
+  defaultValue,
+  value,
+  min = 0,
+  max = HUNDRED,
+  step = 1,
+  disabled,
+  marks,
+  showTicks,
+  showTag,
+  className,
+  renderTag,
+  onChange
+}) => {
+  const ref = useRef<HTMLDivElement>(null);
+  const previouslyDragged = useRef(false);
+  const [values, setValues] = useState(defaultValue ?? min);
+  const validValues: number[] = useMemo(
+    () => toRange(value ?? values, min, max),
+    [max, min, value, values]
+  );
+  const validStep = useMemo(() => (step < 0 ? 0 : step), [step]);
```

Comment:
I would maybe recomment avoid overusing `useMemo` for simple calculations, because it it makes no sense to remember things that are calculated in a single operation

Reply:
You are right, thank you, fixed